### PR TITLE
Test environment and first test cases

### PR DESCRIPTION
**Pull request by ggonzr**
This solves the elements 1, 2, and 3 related to: #1124

Summary (details on each commit message):

1. Build and deploy the McM application based on the branch's code.
2. Configure a minimum sample dataset to run the application. 
3. Include some helper modules for coding the test cases and include constraints on the minimum test cases required.
4. Include a test suite to check the Campaign API.
5. Update the GitHub Action workflow to execute the test suite.

### CODE REVIEW

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -30,6 +30,30 @@ jobs:
...
+          export TEST_FOLDER_PATH=$REPO_PATH/tests
+          python3 -m pip install --upgrade pip
+          python3 -m pip install -r $TEST_FOLDER_PATH/requirements.txt
+
+      - name: Run test suite
         run: |
-          bash deploy/action.sh
\ No newline at end of file
+          export REPO_PATH=$GITHUB_WORKSPACE/repo/mcm
+          export PYTHONPATH=$REPO_PATH
```
Comment (lmoureaux) [inline .github/workflows/test.yml:58]: Side note: we'll want to create a proper package to avoid this hack (poetry?).

File: mcm/deploy/action.sh [hunk 2]
```diff
@@ -7,10 +7,22 @@
...
+if [ -z "$MCM_EXAMPLE_DATA_URL" ]; then
+    echo 'Set $MCM_EXAMPLE_DATA_URL with the URL for downloading the McM data'
+    exit 1
+fi
+
+echo 'Downloading McM data....'
+MCM_DATA_FILE="$HOME/data.tar.gz"
+curl -s -o $MCM_DATA_FILE $MCM_EXAMPLE_DATA_URL
+echo 'Decompressing ....'
+tar -xf $MCM_DATA_FILE -C $HOME/
```
Comment (lmoureaux) [inline mcm/deploy/action.sh:20]: I think this is more idiomatic
```suggestion
curl -s $MCM_EXAMPLE_DATA_URL | tar -xC $HOME
```

**Review by lmoureaux — COMMENTED** [review]
What does the test data contain and how is it obtained? I think this should be documented somewhere.

File: mcm/tests/api_tools.py [hunk 3]
```diff
@@ -0,0 +1,496 @@
...
+            compress_msg = "\n".join(error_msg)
+            raise RuntimeError(compress_msg)
+
+        # Check the required mock users are available
+        # Otherwise, create them
+        if not mockup:
+            if not self.check_test_users():
+                self.__include_test_users()
+
+    def __http_request(
```
Comment (lmoureaux) [inline mcm/tests/api_tools.py:114]: All this method does on top of `requests.request()` is creating the full URL. Remove it.

File: mcm/tests/api_tools.py [hunk 4]
```diff
@@ -0,0 +1,496 @@
...
+            data (dict | None): Data to include in the request body.
+            headers (dict): Request headers
+        """
+        session = Session()
+        full_url: str = base + endpoint
+        req = Request(method=method, url=full_url, headers=headers, json=data)
+        prepared_req = session.prepare_request(req)
+        return session.send(prepared_req)
+
+    def couchdb_request(
```
Comment (lmoureaux) [inline mcm/tests/api_tools.py:138]: This would be a good use case for a lazily created `Session` with `auth=HTTPBasicAuth(...)`. It could then be used with all the normal `requests` methods.

File: mcm/tests/api_tools.py [hunk 5]
```diff
@@ -0,0 +1,496 @@
...
+        full_headers = {**headers, **couchdb_headers}
+        return self.__http_request(
+            base=self.mcm_couchdb_url,
+            endpoint=endpoint,
+            method=method,
+            data=data,
+            headers=full_headers,
+        )
+
+    def mcm_request(
```
Comment (lmoureaux) [inline mcm/tests/api_tools.py:160]: This would be a good use case for a lazily created `Session` with `headers=...` (one per role). It could then be used with all the normal `requests` methods.

File: mcm/tests/api_tools.py [hunk 6]
```diff
@@ -0,0 +1,496 @@
...
+        Sends a HTTP request to McM web application.
+
+        Args:
+            endpoint (str): Endpoint to consume, e.g: entity/document?p=1
+            method (str): HTTP method, e.g: POST
+            as_role (Roles | None): Role to impersonate for the request.
+            data (dict | None): Data to include in the request body.
+            headers (dict): Request headers
+        """
+        full_headers = {**headers}
```
Comment (lmoureaux) [inline mcm/tests/api_tools.py:178]:
```suggestion
        full_headers = headers.copy()
```

File: mcm/tests/api_tools.py [hunk 7]
```diff
@@ -0,0 +1,496 @@
...
+        if as_role:
+            user_headers = {"Adfs-Login": as_role.value}
+            full_headers.update(user_headers)
```
Comment (lmoureaux) [inline mcm/tests/api_tools.py:181]:
```suggestion
            full_headers["Adfs-Login"] = as_role.value
```

File: mcm/tests/api_tools.py [hunk 8]
```diff
@@ -0,0 +1,496 @@
...
+
+        return self.__http_request(
+            base=self.mcm_application_url,
+            endpoint=endpoint,
+            method=method,
+            data=data,
+            headers=full_headers,
+        )
+
+    def lucene_request(
```
Comment (lmoureaux) [inline mcm/tests/api_tools.py:191]: Provide a function to create Lucene URLs instead.

File: mcm/tests/api_tools.py [hunk 9]
```diff
@@ -0,0 +1,496 @@
...
+        new_users: list[dict] = []
+        all_roles = [r.value for r in Roles]
+
+        for role in all_roles:
+            new_user = deepcopy(new_user_mock)
+            new_user["_id"] = role
+            new_user["username"] = role
+            new_user["role"] = role
+            new_user["fullname"] = f"Test user for role: {role}"
+            new_users += [new_user]
```
Comment (lmoureaux) [inline mcm/tests/api_tools.py:249]: Why take the performance hit of creating a new list?
```suggestion
            new_users.append(new_user)
```

File: mcm/tests/api_tools.py [hunk 10]
```diff
@@ -0,0 +1,496 @@
...
+
+        # Send the request.
+        self.couchdb_request(
+            endpoint="users/_bulk_docs", method="POST", data={"docs": new_users}
+        )
```
Comment (lmoureaux) [inline mcm/tests/api_tools.py:254]: I would find it more clear this way:
```suggestion
        self._couchdb_session.post(
            self._couchdb_url("users/_bulk_docs"), json={"docs": new_users}
        )
```

File: mcm/tests/api_tools.py [hunk 11]
```diff
@@ -0,0 +1,496 @@
...
+        response = self.api.mcm_request(
+            endpoint=url, method="PUT", data=data, as_role=self.role
+        )
+        return response.json(), response
+
+    def _delete(self, url) -> tuple[dict, Response]:
+        response = self.api.mcm_request(
+            endpoint=url, method="DELETE", as_role=self.role
+        )
+        return response.json(), response
```
Comment (lmoureaux) [inline mcm/tests/api_tools.py:310]: This is not needed, use native `requests` methods on relevant sessions.

File: mcm/tests/api_tools.py (hunk 4 above)
Comment (lmoureaux) [reply mcm/tests/api_tools.py:138]: If you decide to keep this, pass `*args` and `*kwargs` to `requests.request()`

File: mcm/tests/api_tools.py [hunk 12]
```diff
@@ -0,0 +1,496 @@
...
+    """
+
+    USER = "user"
+    GENERATOR_CONTACT = "generator_contact"
+    GENERATOR_CONVENER = "generator_convener"
+    PRODUCTION_MANAGER = "production_manager"
+    ADMINISTRATOR = "administrator"
+
+
+class APIRequest:
```
Comment (lmoureaux) [inline mcm/tests/api_tools.py:27]: This isn't actually a `Request` in the sense that it's not a derived class of `requests.Request`. So the name is  misleading. I think this could be turned into a bunch of `requests.Session` objects and a few private methods in the `McM` class.

File: mcm/tests/api_tools.py [hunk 13]
```diff
@@ -0,0 +1,496 @@
...
+        return McM(api=self, role=role)
+
+
+class McM:
+    """
+    A simple version for McM REST client adapted to
+    be used for building the test cases.
+
+    Arguments:
+        api (APIRequest): Client to perform HTTP requests.
```
Comment (lmoureaux) [inline mcm/tests/api_tools.py:273]: What is the client here? The `McM` object or the `APIRequest` object?

File: mcm/tests/api_tools.py [hunk 14]
```diff
@@ -0,0 +1,496 @@
...
+        role (Roles): User role to impersonate when performing the
+            requests.
+    """
+
+    def __init__(self, api: APIRequest, role: Roles):
+        self.api: APIRequest = api
+        self.role: Roles = role
+        self.logger = self.__logger()
+
+    def __logger(self) -> logging.Logger:
```
Comment (lmoureaux) [inline mcm/tests/api_tools.py:283]: This function is only used in the constructor, merge it.

File: mcm/tests/api_tools.py [hunk 15]
```diff
@@ -0,0 +1,496 @@
...
+        )
+        return response.json(), response
+
+    # McM methods
+    def get(
+        self, object_type, object_id=None, query="", method="get", page=-1
+    ) -> dict | None:
+        """
+        Get data from McM
+        object_type - [chained_campaigns, chained_requests, campaigns, requests, flows, etc.]
```
Comment (lmoureaux) [inline mcm/tests/api_tools.py:318]: Create an enumeration?

File: mcm/tests/api_tools.py [hunk 16]
```diff
@@ -0,0 +1,496 @@
...
+        object_id - usually prep id of desired object
+        query - query to be run in order to receive an object, e.g. tags=M17p1A, multiple parameters can be used with & tags=M17p1A&pwg=HIG
+        method - action to be performed, such as get, migrate or inspect
+        page - which page to be fetched. -1 means no paginantion, return all results
+        """
+        object_type = object_type.strip()
```
Comment (lmoureaux) [inline mcm/tests/api_tools.py:324]: Remove the `strip()` and just let it fail. Providing it with spaces is a user error.

File: mcm/tests/api_tools.py [hunk 17]
```diff
@@ -0,0 +1,496 @@
...
+        if object_id:
+            object_id = object_id.strip()
```
Comment (lmoureaux) [inline mcm/tests/api_tools.py:326]: Don't strip

File: mcm/tests/api_tools.py [hunk 18]
```diff
@@ -0,0 +1,496 @@
...
+    def _delete(self, url) -> tuple[dict, Response]:
+        response = self.api.mcm_request(
+            endpoint=url, method="DELETE", as_role=self.role
+        )
+        return response.json(), response
+
+    # McM methods
+    def get(
+        self, object_type, object_id=None, query="", method="get", page=-1
+    ) -> dict | None:
```
Comment (lmoureaux) [inline mcm/tests/api_tools.py:315]: Split this in two functions, one for `object_id` and one for `query`. I suggest `get()` and `search()`.

File: mcm/tests/api_tools.py [hunk 19]
```diff
@@ -0,0 +1,496 @@
...
+                method,
+                object_type,
+            )
+            url = "restapi/%s/%s/%s" % (object_type, method, object_id)
+            res, _ = self._get(url)
+            result = res.get("results")
+            if not result:
+                return None
+
+            return result
```
Comment (lmoureaux) [inline mcm/tests/api_tools.py:339]:
```suggestion
            return res.get("results") or None
```

File: mcm/tests/api_tools.py [hunk 20]
```diff
@@ -0,0 +1,496 @@
...
+        return self.put(object_type, object_data, method="update")
+
+    def put(self, object_type, object_data, method="save") -> tuple[dict, Response]:
+        """
+        Put data into McM
+        object_type - [chained_campaigns, chained_requests, campaigns, requests, flows, etc.]
+        object_data - new JSON of an object to be updated
+        method - action to be performed, default is 'save'
+        """
+        url = "restapi/%s/%s" % (object_type, method)
```
Comment (lmoureaux) [inline mcm/tests/api_tools.py:393]:
```suggestion
        url = f"restapi/{object_type}/{method}"
```

File: mcm/tests/api_tools.py [hunk 21]
```diff
@@ -0,0 +1,496 @@
...
+            elif isinstance(root_request, list):
+                # List always contains two elements - start and end of a range
+                query += "%s -> %s\n" % (root_request[0], root_request[1])
+            else:
+                self.logger.error(
+                    "%s is of unsupported type %s", root_request, type(root_request)
+                )
+
+        requests = self.get_range_of_requests(query)
+        return requests
```
Comment (lmoureaux) [inline mcm/tests/api_tools.py:496]: I would simply drop this part of the "API", it does very little.

File: mcm/tests/campaign/test_campaign.py [hunk 22]
```diff
@@ -0,0 +1,266 @@
...
+
+
+class CampaignAPI(Entity):
+    """
+    Implements a simple test API client focused
+    only on the campaign operations.
+    """
+
+    def __init__(self, mcm: McM) -> None:
+        mock_path = Path("./tests/campaign/static/campaign.json").absolute()
```
Comment (lmoureaux) [inline mcm/tests/campaign/test_campaign.py:21]: Use `resource`

File: mcm/tools/locator.py [hunk 23]
```diff
@@ -60,4 +143,75 @@ def database_credentials(self):
...
+
+        Returns:
+            str: Log folder absolute path.
+        """
+        custom_path = os.getenv("MCM_LOG_FOLDER")
+        default_path = os.path.join(os.getcwd(), "logs")
+        if custom_path:
+            self.logger.info("Using a custom log folder: %s", custom_path)
+            return custom_path
+        return default_path
```
Comment (lmoureaux) [inline mcm/tools/locator.py:217]: Only compute `default_path` here if needed

File: mcm/tools/locator.py [hunk 24]
```diff
@@ -1,51 +1,134 @@
...
+    Groups all the required configuration variables for the web application
+    """
+
     def __init__(self):
-        pass
+        self.logger = self.__create_logger()
 
-    def isDev(self):
-        return not (self.isInt() or self.isProd())
+    def __create_logger(self):
```
Comment (lmoureaux) [inline mcm/tools/locator.py:14]: No need for a separate function, do this in `__init__()`

**Review by lmoureaux — CHANGES_REQUESTED** [review]
A good follow-up from this could be a better (structured) system for storing logs.

I'm not convinced by your use of `__functions`. The first time I've seen this used anywhere was in the McM code and name mangling sounds like a niche application. You're not using any complicated inheritance pattern, so there is no need for this. Just use normal private methods.

I haven't looked at the tests themselves yet because there is a lot of support code added in this PR. It's hard to keep track of everything.

**Comment by ggonzr** [comment]
> What does the test data contain and how is it obtained? I think this should be documented somewhere.

The DB dump used on this test is stored in /eos and provided via CERNBox (I will share you and our users the public download URL internally). I created it using as a base the data from the development instance. It includes all the databases but with its content empty. The only exception for this is a database called `settings`. For this database, I left most of the values untouched but I had to censor all the ones related to user e-mails and internal nodes. If this schema (all the databases in general) is not included, the McM application will automatically crash (yeah, unfortunately, it is not possible to test the application without including the DB at the same time, removing this dependency should be a task for the future).

File: .github/workflows/test.yml (hunk 1 above)
Comment (ggonzr) [reply .github/workflows/test.yml:58]: Maybe in the future, we could split different parts of the web application as simple independent Python modules. For the moment, I don’t see the need to use something like `poetry` to package the application.

File: mcm/deploy/action.sh (hunk 2 above)
Comment (ggonzr) [reply mcm/deploy/action.sh:20]: Doesn’t this raise issues in case one chunk of the file is downloaded improperly? Or does this download the complete file first, store it in memory, and then decompress it?

File: mcm/tests/api_tools.py [hunk 25]
```diff
@@ -0,0 +1,496 @@
...
+        Raises:
+            RuntimeError: In case some of the required attributes finish to be empty.
+        """
+        self.mcm_couchdb_url = mcm_couchdb_url or os.getenv("MCM_COUCHDB_URL", "")
+        self.mcm_couchdb_credential = mcm_couchdb_credential or os.getenv(
+            "COUCH_CRED", ""
+        )
+        self.mcm_couchdb_lucene_url = mcm_couchdb_lucene_url or os.getenv(
+            "MCM_LUCENE_URL", ""
+        )
```
[the remark this answers is not in the document]
Comment (ggonzr) [reply mcm/tests/api_tools.py:69]: I want to avoid `None` values as much as possible. That's the reason I prefer to set them to empty strings (also a falsy element) than using `None` directly.

File: mcm/tests/api_tools.py [hunk 26]
```diff
@@ -0,0 +1,496 @@
...
+    endpoints in the McM application impersonating a user or interacting
+    with the database.
+    """
+
+    def __init__(
+        self,
+        mcm_couchdb_url: str = "",
+        mcm_couchdb_credential: str = "",
+        mcm_couchdb_lucene_url: str = "",
+        mcm_application_url: str = "",
```
[the remark this answers is not in the document]
Comment (ggonzr) [reply mcm/tests/api_tools.py:39]: I don't want `None` (nullable values) for native types.

File: mcm/tests/api_tools.py (hunk 12 above)
Comment (ggonzr) [reply mcm/tests/api_tools.py:27]: I agree with this. When I modeled that class, the purpose was to split the responsibilities of consuming endpoints in the databases and the web application and reuse most of the currently available code of the McM REST client. I will merge these tasks in the `McM` class.

File: mcm/tests/api_tools.py (hunk 13 above)
Comment (ggonzr) [reply mcm/tests/api_tools.py:272]: `APIRequest` object

File: mcm/tests/api_tools.py (hunk 14 above)
Comment (ggonzr) [reply mcm/tests/api_tools.py:282]: This is only related to style. I just want to instantiate elements in the constructor and leave its configuration to other methods. It is easier to read

File: mcm/tests/api_tools.py (hunk 15 above)
Comment (ggonzr) [reply mcm/tests/api_tools.py:317]: Agree. I will check in the web application REST API what are all the possible options and pack them in an enum

File: mcm/tests/api_tools.py (hunk 16 above)
Comment (ggonzr) [reply mcm/tests/api_tools.py:323]: Agree. I will remove it

File: mcm/tests/api_tools.py (hunk 17 above)
Comment (ggonzr) [reply mcm/tests/api_tools.py:325]: Agree, it will be removed

File: mcm/tests/api_tools.py (hunk 18 above)
Comment (ggonzr) [reply mcm/tests/api_tools.py:314]: Not sure, but for some tasks and endpoints to test for other entities (`chain_request`, `requests`) would be an advantage to have it like this

File: mcm/tests/api_tools.py [hunk 27]
```diff
@@ -0,0 +1,496 @@
...
+                If it is not provided, it will be tried to retrieve it from
+                `MCM_LUCENE_URL`.
+            mcm_application_url (str): McM web application URL.
+                If it is not provided, it will be tried to retrieve it from
+                - Host: MCM_HOST
+                - Port: MCM_PORT
+            mockup (bool): If True, the creation/validation of mock users
+                will not be perform
+        Raises:
+            RuntimeError: In case some of the required attributes finish to be empty.
```
[the remark this answers is not in the document]
Comment (ggonzr) [reply mcm/tests/api_tools.py:61]: Agree, I will change the exception type

File: mcm/tools/locator.py (hunk 24 above)
Comment (ggonzr) [reply mcm/tools/locator.py:14]: The function `isDev`? If so, I would avoid removing it at this moment. It is used in different places of the McM web application source code. In future iterations, we could scan these kinds of pitfalls and refactor them.

File: mcm/tests/campaign/test_campaign.py (hunk 22 above)
Comment (ggonzr) [reply mcm/tests/campaign/test_campaign.py:21]: I am not sure about what you mean, please provide an example

File: mcm/tests/api_tools.py [hunk 28]
```diff
@@ -0,0 +1,496 @@
...
+        """
+        Instantiate the API request client.
+
+        Attributes:
+            mcm_couchdb_url (str): McM database URL.
+                If it is not provided, it will be tried to retrieve it from
+                `MCM_COUCHDB_URL`.
+            mcm_couchdb_credential (str): Credentials to write data in
+                CouchDB. If it is not provided, it will be tried to retrieve it from
+                `COUCH_CRED`
```
[the remark this answers is not in the document]
Comment (ggonzr) [reply mcm/tests/api_tools.py:50]: [Basic authentication](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Authorization#basic_authentication)

[the remark this answers is not in the document]
Comment (ggonzr) [reply mcm/tests/api_tools.py:50]: Related to the suggested change, the environment is queried only if the parameter was not provided when instantiating the object. Not sure if it is a good idea to keep it like this or maybe it is just better to remove this behavior.

File: mcm/tests/api_tools.py [hunk 29]
```diff
@@ -0,0 +1,496 @@
...
+            mcm_couchdb_lucene_url (str): McM search engine URL.
+                If it is not provided, it will be tried to retrieve it from
+                `MCM_LUCENE_URL`.
```
[the remark this answers is not in the document]
Comment (ggonzr) [reply mcm/tests/api_tools.py:53]: The same as the last comment

File: mcm/tests/api_tools.py [hunk 30]
```diff
@@ -0,0 +1,496 @@
...
+            mcm_application_url (str): McM web application URL.
+                If it is not provided, it will be tried to retrieve it from
+                - Host: MCM_HOST
+                - Port: MCM_PORT
```
[the remark this answers is not in the document]
Comment (ggonzr) [reply mcm/tests/api_tools.py:57]: The variables `MCM_HOST` and `MCM_PORT` are used for the `locator` class to configure the web server. Declaring another variable for the same purposes is redundant.

File: mcm/tools/locator.py (hunk 23 above)
Comment (ggonzr) [reply mcm/tools/locator.py:217]: I think this is more style. I don't see any issue with the current behavior. Closing this comment.

File: mcm/tests/campaign/test_campaign.py (hunk 22 above)
Comment (lmoureaux) [reply mcm/tests/campaign/test_campaign.py:21]:
```py
from importlib.resources import files
import campaign
mock_data = json.loads(files(campaign).joinpath("static/campaign.json").read_text())
```

File: mcm/tests/api_tools.py (hunk 28 above)
[the remark this answers is not in the document]
Comment (lmoureaux) [reply mcm/tests/api_tools.py:50]: > [Basic authentication](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Authorization#basic_authentication)

Sorry, I wanted to say that this should be documented.

> Related to the suggested change, the environment is queried only if the parameter was not provided when instantiating the object. Not sure if it is a good idea to keep it like this or maybe it is just better to remove this behavior.

I'm not sure, I think it would depend on the use cases we foresee. We could also have helper `@classmethod` `from_environment`

File: mcm/tests/api_tools.py (hunk 18 above)
Comment (lmoureaux) [reply mcm/tests/api_tools.py:313]: Not sure how, to pass `kwargs` through?

Comment (lmoureaux) [reply mcm/tests/api_tools.py:313]: Looking at it again, it seems to be `dict | list[dict]`, not `dict | None`?

File: mcm/tools/locator.py (hunk 24 above)
Comment (lmoureaux) [reply mcm/tools/locator.py:14]: No, this was the same as the other comment about `__create_logger`

**Comment by lmoureaux** [comment]
> The DB dump used on this test is stored in /eos and provided via CERNBox (I will share you and our users the public download URL internally). I created it using as a base the data from the development instance. It includes all the databases but with its content empty.

If it doesn't contain any sensitive data, why do we need to keep it private?

File: mcm/tools/locator.py (hunk 24 above)
Comment (ggonzr) [reply mcm/tools/locator.py:14]: Okay, thanks for the clarification. I would like to keep this as it is and the explanation is the same for the others.

File: mcm/tools/locator.py (hunk 23 above)
Comment (lmoureaux) [reply mcm/tools/locator.py:217]: `os.env` is a syscall, which is relatively expensive

Comment (ggonzr) [reply mcm/tools/locator.py:217]: Okay, given that context, I am going to perform the change

File: mcm/tests/campaign/test_campaign.py (hunk 22 above)
Comment (ggonzr) [reply mcm/tests/campaign/test_campaign.py:21]: Nice, I didn't know this. Thanks for the explanation. I am going to perform the change

**Comment by ggonzr** [comment]
> > The DB dump used on this test is stored in /eos and provided via CERNBox (I will share you and our users the public download URL internally). I created it using as a base the data from the development instance. It includes all the databases but with its content empty.
> 
> If it doesn't contain any sensitive data, why do we need to keep it private?

Indeed, it doesn't have any sensitive data but I don't want to expose download links here in GitHub.

File: mcm/deploy/action.sh (hunk 2 above)
Comment (ggonzr) [reply mcm/deploy/action.sh:20]: @lmoureaux...

File: mcm/tests/api_tools.py (hunk 3 above)
Comment (ggonzr) [reply mcm/tests/api_tools.py:113]: Agree, already solved

File: mcm/tests/api_tools.py (hunk 4 above)
Comment (ggonzr) [reply mcm/tests/api_tools.py:137]: This was removed as suggested

File: mcm/tests/api_tools.py (hunk 5 above)
Comment (ggonzr) [reply mcm/tests/api_tools.py:159]: Also removed

File: mcm/tests/api_tools.py (hunk 8 above)
Comment (ggonzr) [reply mcm/tests/api_tools.py:189]: Also refactored

File: mcm/tests/api_tools.py (hunk 10 above)
Comment (ggonzr) [reply mcm/tests/api_tools.py:252]: Already refactored

File: mcm/tests/api_tools.py (hunk 11 above)
Comment (ggonzr) [reply mcm/tests/api_tools.py:308]: I want to avoid too much changes and reuse as much code as possible, I think we can leave this three methods here

File: mcm/tests/api_tools.py (hunk 13 above)
Comment (ggonzr) [reply mcm/tests/api_tools.py:271]: Already refactored, closing this

File: mcm/tests/api_tools.py (hunk 28 above)
[the remark this answers is not in the document]
Comment (ggonzr) [reply mcm/tests/api_tools.py:50]: Let's leave this as already is for the moment

File: mcm/tests/api_tools.py (hunk 12 above)
Comment (ggonzr) [reply mcm/tests/api_tools.py:27]: Already solved, closing this

File: mcm/tests/api_tools.py (hunk 21 above)
Comment (ggonzr) [reply mcm/tests/api_tools.py:490]: Agree, already solved

File: mcm/tests/api_tools.py (hunk 18 above)
Comment (ggonzr) [reply mcm/tests/api_tools.py:313]: Checking again, it is `dict | list[dict] | None`

File: mcm/deploy/action.sh (hunk 2 above)
Comment (lmoureaux) [reply mcm/deploy/action.sh:20]: It's no different than storing the archive on disk and decompressing it later

File: mcm/tests/api_tools.py (hunk 29 above)
[the remark this answers is not in the document]
Comment (lmoureaux) [reply mcm/tests/api_tools.py:53]: This is only about fixing the English

File: mcm/tests/api_tools.py (hunk 30 above)
[the remark this answers is not in the document]
Comment (lmoureaux) [reply mcm/tests/api_tools.py:57]: Fix the English at least

File: mcm/tests/api_tools.py (hunk 29 above)
[the remark this answers is not in the document]
Comment (ggonzr) [reply mcm/tests/api_tools.py]: Solved

File: mcm/tests/api_tools.py (hunk 28 above)
[the remark this answers is not in the document]
Comment (ggonzr) [reply mcm/tests/api_tools.py]: Solved

File: mcm/tests/api_tools.py (hunk 30 above)
[the remark this answers is not in the document]
Comment (ggonzr) [reply mcm/tests/api_tools.py]: Solved

File: mcm/deploy/action.sh (hunk 2 above)
Comment (ggonzr) [reply mcm/deploy/action.sh:20]: Okay, let's include it

File: mcm/tests/campaign/test_campaign.py [hunk 31]
```diff
@@ -0,0 +1,268 @@
...
+"""
+
+import uuid
+from pathlib import Path
+from copy import deepcopy
+from importlib.resources import files
+import pytest
+import tests.campaign as campaign
+from tests.base_test_tools import Entity, EntityTest, config
+from tests.api_tools import McM, Roles
```
Comment (lmoureaux) [inline mcm/tests/campaign/test_campaign.py:13]: I recommend running `isort` on this file

File: mcm/tests/base_test_tools.py [hunk 32]
```diff
@@ -0,0 +1,141 @@
...
+    mcm_couchdb_lucene_url="http://localhost:5985/",
+    mcm_application_url="http://localhost:8000/",
+)
+
+
+class Entity(abc.ABC):
+    """
+    Minimum common operations to test the API related to
+    one McM entity/concept, e.g: campaigns, requests,
+    chain_requests, etc.'
```
Comment (lmoureaux) [inline mcm/tests/base_test_tools.py:27]: Typo
```suggestion
    chain_requests, etc.
```

File: mcm/tests/campaign/test_campaign.py [hunk 33]
```diff
@@ -0,0 +1,268 @@
...
+
+    def delete(self, mockup: dict):
+        return self.mcm.delete(self.object_type, mockup.get("prepid"))
+
+    def example(self):
+        mockup = self.mockup()
+        self.create(mockup)
+        return self.retrieve(mockup)
+
+    def cms_drivers(self, mockup: dict):
```
Comment (lmoureaux) [inline mcm/tests/campaign/test_campaign.py:52]: What's the return type of this function?

File: mcm/tests/campaign/test_campaign.py [hunk 34]
```diff
@@ -0,0 +1,268 @@
...
+        """
+        Retrieves the cms-sw commands for a given campaign.
+        Related to test the endpoint:
+            - restapi/campaigns/get_cmsDrivers/<string:campaign_id>
+        """
+        endpoint = f"restapi/campaigns/get_cmsDrivers/{mockup.get('prepid')}"
+        return self.mcm._get(endpoint)
+
+    def inspect(self, mockup: dict):
```
Comment (lmoureaux) [inline mcm/tests/campaign/test_campaign.py:61]: What's the return type of this function?

File: mcm/tests/campaign/test_campaign.py [hunk 35]
```diff
@@ -0,0 +1,268 @@
...
+        """
+        Inspect all the requests in a given campaign.
+        Related to test the endpoint:
+            - restapi/campaigns/inspect/<string:campaign_id>
+        """
+        endpoint = f"restapi/campaigns/inspect/{mockup.get('prepid')}"
+        return self.mcm._get(endpoint)
+
+    def status(self, mockup: dict):
```
Comment (lmoureaux) [inline mcm/tests/campaign/test_campaign.py:70]: * Maybe `toggle_status` since it changes state?
* What does it return?

File: mcm/tests/campaign/test_campaign.py [hunk 36]
```diff
@@ -0,0 +1,268 @@
...
+
+        # Update a campaign that doesn't exist.
+        example = self.entity_api.example()
+        example["prepid"] = "Run3Campaign#2"
+        content, response = self.entity_api.update(example)
+
+        assert response.status_code == 200
+        assert content["results"] == False
+
+    @pytest.mark.skip("Takes too much time to compute")
```
Comment (lmoureaux) [inline mcm/tests/campaign/test_campaign.py:141]: How comes it's so slow? I think the test should be implemented but marked as [`slow`](https://docs.pytest.org/en/7.1.x/how-to/mark.html#mark).

File: mcm/tests/campaign/test_campaign.py [hunk 37]
```diff
@@ -0,0 +1,268 @@
...
+        assert driver.startswith("cmsDriver.py")
+        assert event_content_param in driver
+
+        # The campaign doesn't exists
+        mockup = self.entity_api.mockup()
+        content, response = self.entity_api.cms_drivers(mockup)
+        assert response.status_code == 200
+        assert content["results"] == False
+
+    @pytest.mark.skip("Takes too much time to computes and renders a html page")
```
Comment (lmoureaux) [inline mcm/tests/campaign/test_campaign.py:162]: What makes it so slow in a mock database with just a few records? I think it should be implemented but marked as [`slow`](https://docs.pytest.org/en/7.1.x/how-to/mark.html#mark). For HTML we could test `Content-Type` (if it's set correctly...) or use `html.parser`

**Review by lmoureaux — CHANGES_REQUESTED** [review]
I now went through the test code.

As we discussed on Mattermost, I think we shouldn't use `__functions` unless necessary. Python has a more open approach than Java regarding code reuse (someone using a `_function` is supposed to know what they are doing and it's the caller's responsibility to ensure that they don't break anything)

File: mcm/tests/campaign/test_campaign.py [hunk 38]
```diff
@@ -0,0 +1,268 @@
...
+        return self.mcm.put(self.object_type, mockup)
+
+    def retrieve(self, mockup: dict):
+        return self.mcm.get(self.object_type, mockup.get("prepid"))
+
+    def update(self, mockup: dict):
+        return self.mcm.update(self.object_type, mockup)
+
+    def delete(self, mockup: dict):
+        return self.mcm.delete(self.object_type, mockup.get("prepid"))
```
[the remark this answers is not in the document]
Comment (ggonzr) [reply mcm/tests/campaign/test_campaign.py:45]: I understand your point of view, but maybe future entities like `chain_campaign` and `chain_request` would require more complex procedures, so let's leave the child class to implement it. In the end, we can give it a refactor in case it is required.

File: mcm/tests/campaign/test_campaign.py (hunk 33 above)
Comment (ggonzr) [reply mcm/tests/campaign/test_campaign.py:52]: A `dict` with a key called `results`  wrapping a `list[dict]` describing the cms-sw command for simulating each step.

File: mcm/tests/campaign/test_campaign.py (hunk 34 above)
Comment (ggonzr) [reply mcm/tests/campaign/test_campaign.py:61]: A `dict` with a bool field called `result` represents whether the operation finished successfully or not and some fields describing error messages could appear.

File: mcm/tests/campaign/test_campaign.py (hunk 35 above)
Comment (lmoureaux) [reply mcm/tests/campaign/test_campaign.py:70]: Let's document this in the docstring and add `-> dict`?

File: mcm/tests/campaign/test_campaign.py (hunk 36 above)
Comment (ggonzr) [reply mcm/tests/campaign/test_campaign.py:141]: I have included a test, need to check if works properly when deploying into test env

**Review by lmoureaux — APPROVED** [review]
